### PR TITLE
CDev begin extracting from drivers Device hierarchy

### PR DESCRIPTION
**Pull request by dagar**
CDev is inserted in the middle of the Device interface between Device and I2C or SPI. As a result there are many things in the system using CDev that unnecessarily carry Device.

This includes...
 - all uORB topics and the uORB manager
 - all "drivers", including those that don't use I2C or SPI

The Device base class includes the uint32_t deviceid, a debugging bool (_debug_enabled) and a string that's only used in DEVICE_DEBUG and DEVICE_LOG helpers.

Overall this **saves at least a couple kilobytes of RAM** on a typical setup, but there are still many more cases of CDev usage in the system where only a CDev was needed without Device, or where a Device was needed without CDev. There are also many situations where most of the CDevs aren't needed at all (eg /dev/mpu6000_accel, /dev/icm20602_accel, /dev/icm20608_accel, etc).

This is a work in progress that isn't finished yet. Ideally I'd like to extract CDev 100%, but that would require minor changes in nearly every driver.

### CODE REVIEW

**Comment by dagar** [comment]
Key changes
 - cdev is a new standalone library
 - drivers/cdev is now a shim enabling deviceid access via cdev ioctl (once drivers are updated to access the device id directly this can be killed)
 - uORB (all topics), px4fmu, px4io, and a few others are now using the standalone cdev directly, instead of cdev with the Driver base class.

**Comment by bkueng** [comment]
Nice!
3 initial remarks:
- Do you plan to rename one of the 2 `CDev` classes? The updated structure adds quite a bit of confusion.
- `_pub_blocked` is actually useful for testing sensor failover for example (I'm ok if we have some other method to achieve that)
- Can you split into separate commits? It will make review and memory saving assessments easier.

**Comment by dagar** [comment]
> Do you plan to rename one of the 2 CDev classes? The updated structure adds quite a bit of confusion.

How about CDevAdaptor for what remains under Device? The goal is to have this removed entirely soon.

> _pub_blocked is actually useful for testing sensor failover for example (I'm ok if we have some other method to achieve that)

Yes I like it in theory, but it's one of those things that's implemented inconsistently. For example `config block /dev/accel0` will block the mpu6000's accel and gyro orb publications, because the main MPU6000 object is the accel CDev. Maybe we should think about a mechanism directly in uORB?

> Can you split into separate commits? It will make review and memory saving assessments easier.

I'll see if I can split it a bit, but ultimately I want every commit to be fully functional.

**Comment by dagar** [comment]
> > Do you plan to rename one of the 2 CDev classes? The updated structure adds quite a bit of confusion.
> 
> How about CDevAdaptor for what remains under Device? The goal is to have this removed entirely soon.

On second thought this will probably make an even bigger mess. Given how convoluted it already is I'd rather push for removing it entirely. The only reason this can't be done trivially is that some drivers (not all) rely on DEVIOCGDEVICEID ioctl calls going through to CDev::ioctl.

**Comment by dagar** [comment]
Rebased on master.

**Comment by dagar** [comment]
Next steps
1. manually move all remaining usage over to new standalone cdev
2. delete Device CDev shim

**Comment by santiago3dr** [comment]
Couple flights on Pixracer (V4)

good flights; no noticeable difference from master
https://logs.px4.io/plot_app?log=573371f3-4d7a-4a2a-b4d2-c7fd0d9aad73
https://logs.px4.io/plot_app?log=b51dbc99-4a7e-4314-8fda-64ebdb451c95

**Comment by Avysuarez** [comment]
Couple flights with pixhawk 4 (v5)
good flights (Pr)
https://review.px4.io/plot_app?log=216d5a21-7954-4c75-9d60-908aa75e3ea5
https://review.px4.io/plot_app?log=712aa787-bf77-4bf6-9058-4dd72ec12809

good flights. (master)
https://review.px4.io/plot_app?log=3b519f70-b00e-48e1-ae8e-12390b41154c
https://review.px4.io/plot_app?log=9acaca64-cd71-417d-bc59-03a277d1fa8d

**Comment by dannyfpv** [comment]
Couple flights on Pixhawk v2
https://review.px4.io/plot_app?log=ccf77c90-84e7-4132-839d-1e3b9d71cfbb
https://review.px4.io/plot_app?log=283a8ebc-e465-4a05-b695-6c3e3f54c15c

couple flights on pixhawk mini v3
https://review.px4.io/plot_app?log=911ca722-84d8-4c95-8947-7136fa182abd
https://review.px4.io/plot_app?log=9e44fa94-a450-4f6c-b56b-86f458a57c9f

couple flights on pixhawk pro v4
https://review.px4.io/plot_app?log=99482a25-653c-49d4-8831-ec19e193bd2d
https://review.px4.io/plot_app?log=4e45d2e8-3634-4b36-bdfa-631b3161bde1

**Comment by dagar** [comment]
@bkueng @davids5 any comment? I'd like to get this in and keep picking away at cdev abuse until it can be completely removed from the Device hierarchy.

File: src/lib/cdev/CDev.hpp [hunk 1]
```diff
@@ -0,0 +1,300 @@
...
+
+private:
+	const char	*_devname{nullptr};		/**< device node name */
+
+	px4_pollfd_struct_t	**_pollset{nullptr};
+
+	bool		_registered{false};		/**< true if device name was registered */
+
+	uint8_t		_max_pollwaiters{0};		/**< size of the _pollset array */
+	uint8_t		_open_count{0};			/**< number of successful opens */
```
Comment (bkueng) [inline src/lib/cdev/CDev.hpp]: We can keep this uint16_t (just in case), it should not increase memory usage.

Comment (dagar) [reply src/lib/cdev/CDev.hpp]: Done

**Comment by dagar** [comment]
For reference to complete the transition we need to replace all `device::CDev` with `cdev::CDev` and then any cases where IOCTL calls are made directly to a `device::I2C` or `device::SPI` need to inherit `cdev::CDev`.

The majority of drivers have a frontend piece (eg mpu6000) that should be a CDev (/dev/accel0, /dev/gyro0), but *not* a Device (a bus, address, and type). IOCTL device id calls can be passed to the particular I2C or SPI interface (which is a Device).

This type of thing is usually a sign of confusion.
https://github.com/PX4/Firmware/blob/master/src/drivers/imu/mpu6000/mpu6000.cpp#L515-L518

**Pull request merged**